Content.Server/DeviceLinking/Systems/ObjectSensorSystem.cs

```diff
@@ -0,0 +1,151 @@
+    [Dependency] private readonly DeviceLinkSystem _deviceLink = default!;
+    [Dependency] private readonly UseDelaySystem _useDelay = default!;
+
+    private readonly int ModeCount = Enum.GetValues(typeof(ObjectSensorMode)).Length;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<ObjectSensorComponent, ComponentInit>(OnInit);
+        SubscribeLocalEvent<ObjectSensorComponent, ExaminedEvent>(OnExamined);
+        SubscribeLocalEvent<ObjectSensorComponent, InteractUsingEvent>(OnInteractUsing);
+    }
+
+    public override void Update(float deltaTime)
+    {
+        var query = EntityQueryEnumerator<ObjectSensorComponent>();
+        while (query.MoveNext(out var uid, out var component))
+        {
+            UpdateOutput(uid, component);
+        }
+    }
+
+    private void OnInteractUsing(EntityUid uid, ObjectSensorComponent component, InteractUsingEvent args)
+    {
+        if (args.Handled || !_tool.HasQuality(args.Used, component.CycleQuality))
+            return;
+
+        if (TryComp<UseDelayComponent>(uid, out var delay)
+            && !_useDelay.TryResetDelay((uid, delay), true))
+            return;
+
+        var mode = (int) component.Mode;
+        mode = ++mode % ModeCount;
+        component.Mode = (ObjectSensorMode) mode;
+
+        UpdateOutput(uid, component);
+
+        _audio.PlayPvs(component.CycleSound, uid);
+        var msg = Loc.GetString("object-sensor-cycle", ("mode", component.Mode.ToString()));
+        _popup.PopupEntity(msg, uid, args.User);
+    }
+
+    private void OnExamined(EntityUid uid, ObjectSensorComponent component, ExaminedEvent args)
+    {
+        if (!args.IsInDetailsRange)
+            return;
+
+        args.PushMarkup(Loc.GetString("object-sensor-examine", ("mode", component.Mode.ToString())));
+    }
+
+    private void OnInit(EntityUid uid, ObjectSensorComponent component, ComponentInit args)
+    {
+        _deviceLink.EnsureSourcePorts(uid, component.OutputPort1, component.OutputPort2, component.OutputPort3, component.OutputPort4OrMore);
+    }
+
+    private void SetOut(EntityUid uid, ObjectSensorComponent component, int port, bool signal)
+    {
+        Logger.Debug($"{ToPrettyString(uid)} {port} {signal}");
+        switch (port)
+        {
+            case 1:
+                _deviceLink.SendSignal(uid, component.OutputPort1, signal);
+                break;
+            case 2:
+                _deviceLink.SendSignal(uid, component.OutputPort2, signal);
+                break;
+            case 3:
+                _deviceLink.SendSignal(uid, component.OutputPort3, signal);
+                break;
+            case 4:
+                _deviceLink.SendSignal(uid, component.OutputPort4OrMore, signal);
+                break;
+        }
+    }
+
+    private void UpdateOutput(EntityUid uid, ObjectSensorComponent? component)
+    {
+        if (!Resolve(uid, ref component))
```

Comment:
make it non optional and remove resolve since everything has the component to pass